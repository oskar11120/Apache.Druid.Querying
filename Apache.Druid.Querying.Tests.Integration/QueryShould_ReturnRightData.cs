using FluentAssertions;
using Snapshooter;
using Snapshooter.NUnit;
using System.Globalization;
using System.Runtime.CompilerServices;
using static Apache.Druid.Querying.Tests.Integration.ServiceProvider;

namespace Apache.Druid.Querying.Tests.Integration;

internal static class TestData
{
    public static readonly DateTimeOffset T0 = DateTimeOffset.Parse("2016-06-27T00:00:00.000Z", CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal);
    public static readonly Interval Interval = new(T0, T0.AddDays(1));

    public static TQuery DefaultInterval<TQuery>(this TQuery query)
        where TQuery : IQueryWith.Intervals
        => query
            .Interval(Interval);
}

internal class QueryShould_ReturnRightData
{
    private static async Task VerifyMatch<TSource, TResult, TContext>(
        DataSource<TSource> dataSource,
        IQueryWithSource<TSource>.AndResult<TResult>.AndDeserializationAndTruncatedResultHandling<TContext> query,
        [CallerArgumentExpression(nameof(query))] string snapshotNameSuffix = "")
        where TContext : new()
    {
        var spanshotName = Snapshot.FullName(new SnapshotNameExtension(snapshotNameSuffix));
        var json = dataSource.MapQueryToJson(query).ToString();
        await TestContext.Out.WriteLineAsync(json);
        void Match(List<TResult> results)
            => Snapshot.Match(results, spanshotName, options => options.IgnoreField("[:].SegmentId"));
        var results = await dataSource
            .ExecuteQuery(query)
            .ToListAsync();
        Match(results);

        var withNoTruncatedResultHandling = await dataSource
            .ExecuteQuery(query, onTruncatedResultsQueryRemaining: false)
            .ToListAsync();
        Match(withNoTruncatedResultHandling);
    }

    private static Task VerifyMatch<TResult, TContext>(
        IQueryWithSource<Edit>.AndResult<TResult>.AndDeserializationAndTruncatedResultHandling<TContext> query,
        [CallerArgumentExpression(nameof(query))] string snapshotNameSuffix = "")
        where TContext : new()
        => VerifyMatch(Wikipedia.Edits, query, snapshotNameSuffix);

    [Test]
    public async Task Wrapped()
    {
        var first = new Query<Edit>
            .TimeSeries
            .WithNoVirtualColumns
            .WithAggregations<TimeSeriesAggregations>()
            .Aggregations(type => new(
                type.Count(),
                type.Sum(edit => edit.Added)))
            .DefaultInterval()
            .Granularity(Granularity.FifteenMinutes)
            .Context(new() { SkipEmptyBuckets = true });
        var wrapped = Wikipedia
            .Edits
            .ToQueryDataSource(first);
        var query = new Query<WithTimestamp<TimeSeriesAggregations>>
            .Scan()
            .Interval(TestData.Interval)
            .Filter(type => type.Not(type.Selector(data => data.Value.Count, 6)))
            .Limit(1000);
        await VerifyMatch(wrapped, query, string.Empty);
    }

    public enum JoinTestCase
    {
        LatestEditTimestampsPerCountry,
        LatestEditsPerCountry,
        Both
    }
    private record Country(string Code, string FullName);
    [TestCase(JoinTestCase.LatestEditTimestampsPerCountry)]
    [TestCase(JoinTestCase.LatestEditsPerCountry)]
    [TestCase(JoinTestCase.Both)]
    public async Task Join(JoinTestCase @case)
    {
        var latestEditTimestampsPerCountry_query = new Query<Edit>
            .GroupBy<string>
            .WithNoVirtualColumns
            .WithAggregations<DateTimeOffset>()
            .DefaultInterval()
            .Granularity(Granularity.Day)
            .Dimensions(type => type.Default(data => data.CountryName))
            .Aggregations(type => type.Last(data => data.Timestamp));
        var latestEditTimestampsPerCountry_dataSource = Wikipedia.Edits.ToQueryDataSource(latestEditTimestampsPerCountry_query);
        var lastestEditsPerCountry_dataSource = Wikipedia.Edits.InnerJoin(
            latestEditTimestampsPerCountry_dataSource,
            data => $"{data.Left.CountryName} == {data.Right.Value.Dimensions} && {data.Left.Timestamp} == {data.Right.Value.Aggregations}");
        var edits_query = new Query<InnerJoinData<Edit, WithTimestamp<Dimensions_Aggregations<string, DateTimeOffset>>>>
            .Scan()
            .DefaultInterval();
        await (@case switch
        {
            JoinTestCase.LatestEditTimestampsPerCountry => VerifyMatch(Wikipedia.Edits, latestEditTimestampsPerCountry_query, @case.ToString()),
            JoinTestCase.LatestEditsPerCountry => VerifyMatch(lastestEditsPerCountry_dataSource, edits_query, @case.ToString()),
            JoinTestCase.Both => VerifyBoth(),
            _ => throw new NotSupportedException()
        });

        async Task VerifyBoth()
        {
            var fromEdits = await lastestEditsPerCountry_dataSource
                .ExecuteQuery(edits_query)
                .Select(result => result.Value.Right.Value)
                .OrderBy(result => result.Aggregations)
                .ToArrayAsync();
            var fromTimestamps = await Wikipedia.Edits
                .ExecuteQuery(latestEditTimestampsPerCountry_query)
                .Select(result => result.Value)
                .Where(result => result.Dimensions is not null)
                .OrderBy(result => result.Aggregations)
                .ToArrayAsync();
            fromEdits.Should().BeEquivalentTo(fromTimestamps);
        }
    }

    [Test]
    public async Task Union()
    {
        var union = Wikipedia
            .Edits
            .Union(Wikipedia
                .Edits);
        var query = new Query<Union<Edit, Edit>>
            .Scan()
            .Interval(TestData.Interval)
            .Limit(10);
        await VerifyMatch(union, query, string.Empty);
    }

    private sealed record InlineData(string Word, [property: DataSourceColumn("Integer")] int Number);
    [Test]
    public async Task Inline()
    {
        var query = new Query<InlineData>
            .Scan()
            .Interval(new(default, DateTimeOffset.MaxValue))
            .Filter(type => type.In(
                data => data.Word,
                new[] { "zero" }));
        var inline = Wikipedia
            .Inline(new InlineData[]
            {
                new("zero", 1),
                new("one", 2)
            });
        var result = await inline
            .ExecuteQuery(query)
            .ToListAsync();
        result
            .Single()
            .Value
            .Should()
            .Be(new InlineData("zero", 1));
    }

    [Test]
    public async Task Scan()
    {
        var first3Us = new Query<Edit>
            .Scan()
            .DefaultInterval()
            .Filter(type => type.Selector(
                edit => edit.CountryIsoCode,
                "US"))
            .Limit(3);
        await VerifyMatch(first3Us);
    }

    [Test]
    public async Task GroupBy()
    {
        var countsPerCountry = new Query<Edit>
            .GroupBy<string>
            .WithNoVirtualColumns
            .WithAggregations<int>()
            .DefaultInterval()
            .Dimensions(type => type.Default(edit => edit.CountryName))
            .Aggregations(type => type.Count())
            .Granularity(Granularity.EightHours);

        var countsPerCountryOrderedByCount = countsPerCountry
            .LimitSpec(limit: 100, offset: 1, columns => columns.OrderBy(data => data.Aggregations, OrderDirection.Descending, SortingOrder.Numeric));
        await VerifyMatch(countsPerCountryOrderedByCount);

        var countsPerValidCountry = countsPerCountry
            .Filter(type => type.Not(type.Selector(edit => edit.CountryName, string.Empty)));
        await VerifyMatch(countsPerValidCountry);
    }

    [Test]
    public async Task TopN()
    {
        var countsPerPage = new Query<Edit>
            .TopN<string>
            .WithNoVirtualColumns
            .WithAggregations<int>()
            .Dimension(type => type.Default(edit => edit.Page))
            .Threshold(10)
            .Aggregations(type => type.Count())
            .Metric(type => type.Numeric(data => data.Aggregations))
            .DefaultInterval()
            .Granularity(Granularity.Hour);
        await VerifyMatch(countsPerPage);

        var usAnonymousCountsPerPage = countsPerPage
            .Filter(type => type.And(
                type.Selector(edit => edit.IsAnonymous, true),
                type.Selector(edit => edit.CountryIsoCode, "US")));
        await VerifyMatch(usAnonymousCountsPerPage);
    }

    private record TimeSeriesAggregations(int Count, int TotalAdded);
    private record TimeSeriesPostAggrgations(double AverageAdded);
    [Test]
    public async Task TimeSeries()
    {
        var lineStatisticsPerHour = new Query<Edit>
            .TimeSeries
            .WithNoVirtualColumns
            .WithAggregations<TimeSeriesAggregations>
            .WithPostAggregations<TimeSeriesPostAggrgations>()
            .Order(OrderDirection.Descending)
            .Aggregations(type => new(
                type.Count(),
                type.Sum(edit => edit.Added)))
            .PostAggregations(type => new(type.Arithmetic(
                ArithmeticFunction.Divide,
                type.FieldAccess(aggregations => aggregations.TotalAdded),
                type.FieldAccess(aggregations => aggregations.Count))))
            .Filter(type => type.Selector(edit => edit.CountryIsoCode, "US"))
            .DefaultInterval()
            .Granularity(Granularity.Hour)
            .Context(new QueryContext.TimeSeries() { SkipEmptyBuckets = true });
        await VerifyMatch(lineStatisticsPerHour);

        var inEuWarsawTime = lineStatisticsPerHour
            .Granularity(Granularity.Hour, "Europe/Warsaw");
        await VerifyMatch(inEuWarsawTime);
    }


    private sealed record GroupByBooleans(bool Robot, bool New);
    [TestCase(0)]
    [TestCase(1)]
    public async Task WithDataPropertiesAccessedByInterface(int i)
    {
        async Task Test<TEdit>(DataSource<TEdit> dataSource) where TEdit : IEditBooleans
        {
            var one = new Query<TEdit>
                .Scan()
                .DefaultInterval()
                .Filter(type => type.And(
                    type.Selector(data => data.IsNew, true),
                    type.Selector(data => data.Robot, true)))
                .Limit(100);
            if (i is 0)
                await VerifyMatch(dataSource, one);

            var two = new Query<TEdit>
                .GroupBy<GroupByBooleans>
                .WithNoVirtualColumns
                .WithAggregations<int>()
                .DefaultInterval()
                .Granularity(Granularity.Hour)
                .Dimensions(type => new(
                    type.Default(data => data.Robot),
                    type.Default(data => data.IsNew)))
                .Aggregations(type => type.Count());
            if (i is 1)
                await VerifyMatch(dataSource, two);
        }

        await Test(Wikipedia.Edits);
    }

    interface IIotMeasurement
    {
        Guid ObjectId { get; }
        double Value { get; }
    }
}
