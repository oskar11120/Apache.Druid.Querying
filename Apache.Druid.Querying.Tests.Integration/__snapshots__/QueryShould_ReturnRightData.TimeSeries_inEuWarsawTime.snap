[
  {
    "Timestamp": "2016-06-27T23:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 17,
        "TotalAdded": 1260
      },
      "PostAggregations": {
        "AverageAdded": 74.11764705882354
      }
    }
  },
  {
    "Timestamp": "2016-06-27T22:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 25,
        "TotalAdded": 2150
      },
      "PostAggregations": {
        "AverageAdded": 86.0
      }
    }
  },
  {
    "Timestamp": "2016-06-27T21:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 27,
        "TotalAdded": 10380
      },
      "PostAggregations": {
        "AverageAdded": 384.44444444444446
      }
    }
  },
  {
    "Timestamp": "2016-06-27T20:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 26,
        "TotalAdded": 2430
      },
      "PostAggregations": {
        "AverageAdded": 93.46153846153847
      }
    }
  },
  {
    "Timestamp": "2016-06-27T19:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 25,
        "TotalAdded": 1622
      },
      "PostAggregations": {
        "AverageAdded": 64.88
      }
    }
  },
  {
    "Timestamp": "2016-06-27T18:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 19,
        "TotalAdded": 9819
      },
      "PostAggregations": {
        "AverageAdded": 516.7894736842105
      }
    }
  },
  {
    "Timestamp": "2016-06-27T17:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 24,
        "TotalAdded": 1789
      },
      "PostAggregations": {
        "AverageAdded": 74.54166666666667
      }
    }
  },
  {
    "Timestamp": "2016-06-27T16:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 20,
        "TotalAdded": 1031
      },
      "PostAggregations": {
        "AverageAdded": 51.55
      }
    }
  },
  {
    "Timestamp": "2016-06-27T15:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 16,
        "TotalAdded": 1733
      },
      "PostAggregations": {
        "AverageAdded": 108.3125
      }
    }
  },
  {
    "Timestamp": "2016-06-27T14:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 7,
        "TotalAdded": 620
      },
      "PostAggregations": {
        "AverageAdded": 88.57142857142857
      }
    }
  },
  {
    "Timestamp": "2016-06-27T13:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 6,
        "TotalAdded": 620
      },
      "PostAggregations": {
        "AverageAdded": 103.33333333333333
      }
    }
  },
  {
    "Timestamp": "2016-06-27T12:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 7,
        "TotalAdded": 204
      },
      "PostAggregations": {
        "AverageAdded": 29.142857142857142
      }
    }
  },
  {
    "Timestamp": "2016-06-27T11:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 4,
        "TotalAdded": 467
      },
      "PostAggregations": {
        "AverageAdded": 116.75
      }
    }
  },
  {
    "Timestamp": "2016-06-27T10:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 5,
        "TotalAdded": 6
      },
      "PostAggregations": {
        "AverageAdded": 1.2
      }
    }
  },
  {
    "Timestamp": "2016-06-27T09:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 4,
        "TotalAdded": 407
      },
      "PostAggregations": {
        "AverageAdded": 101.75
      }
    }
  },
  {
    "Timestamp": "2016-06-27T08:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 6,
        "TotalAdded": 73
      },
      "PostAggregations": {
        "AverageAdded": 12.166666666666666
      }
    }
  },
  {
    "Timestamp": "2016-06-27T07:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 9,
        "TotalAdded": 218
      },
      "PostAggregations": {
        "AverageAdded": 24.22222222222222
      }
    }
  },
  {
    "Timestamp": "2016-06-27T06:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 20,
        "TotalAdded": 746
      },
      "PostAggregations": {
        "AverageAdded": 37.3
      }
    }
  },
  {
    "Timestamp": "2016-06-27T05:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 20,
        "TotalAdded": 2354
      },
      "PostAggregations": {
        "AverageAdded": 117.7
      }
    }
  },
  {
    "Timestamp": "2016-06-27T04:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 13,
        "TotalAdded": 301
      },
      "PostAggregations": {
        "AverageAdded": 23.153846153846153
      }
    }
  },
  {
    "Timestamp": "2016-06-27T03:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 23,
        "TotalAdded": 268
      },
      "PostAggregations": {
        "AverageAdded": 11.652173913043478
      }
    }
  },
  {
    "Timestamp": "2016-06-27T02:00:00+02:00",
    "Value": {
      "Aggregations": {
        "Count": 16,
        "TotalAdded": 1423
      },
      "PostAggregations": {
        "AverageAdded": 88.9375
      }
    }
  }
]
